Dbz 175: In-line Snapshots of new tables

mtagle opened the pull request:
This is by no means complete at this point.  However, I've done a pretty significant refactor job of snapshots and I want to have some extra eyes that are more familiar with the snapshotting process take a look and see what incorrect assumptions I'm making about the snapshot code/process/order.

Of course, any other feedback or questions are very welcome!

xenji commented:
As discussed with Moira in a gitter chat, we should think about the event of a binlog getting rotated away while a snapshot still runs. This would mean, that we cannot jump back to the binlog position that was recorded before the snapshot started. 

Would this mean we need to re-snapshot everything? Does the current initial snapshot process have the same risk?

rhauch commented:
There is a difference between binlog rotation, where the server stops writing to the current binlog file starts writing to a new binlog file, and removal of binlog files after they "expire". The former is a normal part of the binlog mechanism, and the current code handles this well: it tracks the binlog file and position in that file in every offset, and updates the binlog file on a binlog rotation event. Upon restart, it can resume reading from the binlog by starting at the position in the specified binlog file _as long as that binlog file is still around_. If a binlog file is not available, the connector will perform a new snapshot (if snapshot mode is "as needed") or willl fail.

xenji commented:
Ok just to summarize in my words: The 10th binlog file (let's say I have 10 configured) gets rotated away and contains the offset means re-snapshot-ing everything. If the binlog position is later than the earliest position in the 10th file, everything is good.

jpechane commented on debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/SnapshotServerReader.java:
```diff
@@ -0,0 +1,216 @@
+                if (!isRunning()) return;
+                //logger.info("\tsnapshot continuing with database(s): {}", includedDatabaseNames);
+
+                snapshotMySQLUtility.maybeReadTableList();
```
This is slightly confusing - it should be clear that readTableList was executed as a part of lockAllTables whel global locak failed. How about passing two lambdas - one executed when global lock succeeds and the other one when it fails?

mtagle replied:
Changed to explicitly read the table list iff the global lock succeed, and some comments about why.

jpechane commented on debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/SnapshotServerReader.java:
```diff
@@ -0,0 +1,216 @@
+        try {
+            metrics.startSnapshot();
+
+            snapshotMySQLUtility.beginTransaction();
```
Make sure that changes in https://github.com/debezium/debezium/pull/274 are incroporated

mtagle replied:
ty, fixed.

jpechane commented on debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/SnapshotMySQLUtility.java:
```diff
@@ -0,0 +1,543 @@
+    /**
+     * Complete the current transaction, and unlock any tables that are still locked.
+     */
+    public void completeTransaction() throws SQLException {
```
Please consider renaming - you are completing TX AND releasing the table locks

mtagle replied:
renamed

jpechane commented on debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/SnapshotServerReader.java:
```diff
@@ -0,0 +1,216 @@
+                // From this point forward, all source records produced by this connector will have an offset that includes a
+                // "snapshot" field (with value of "true").
+
+                // ------
```
I like the idea of the flow to be split into method calls but I think it should be done for all steps/operations. As this code iw very complex and it takes time to understand it would be great if we would be able to describe MySQL snapshotting process with high-level names/block and ALL those blocks would be transformed into methods so the execute method will turn into a self-documenting workflow as it will only call those blocks and do just meybe some decision making.

mtagle replied:
Fair enough, I'm also a fan of this complex logic being more readable.  I'll see what other break-downs I can do.

jpechane commented on debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlConnectorTask.java:
```diff
@@ -195,6 +201,42 @@ public synchronized void start(Map<String, String> props) {
+
+            Set<TableId> newKnownTables = new HashSet<>();
+            for (TableId tableId : recordMakers.getAllTableIds()) {
+                if (!oldFilters.tableFilter().test(tableId) && newFilters.tableFilter().test(tableId)){
```
How this will work when there is a new table created during connector downtime that was already covered by old filter?

mtagle replied:
I changed this logic a bit, so now I'm only determining what needs to be snapshotted from the config, at startup. If a new table is added during connector downtime it should not be snapshotted if it was covered by the old filter.

mtagle commented on debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlConnectorTask.java:
```diff
@@ -195,6 +201,42 @@ public synchronized void start(Map<String, String> props) {
+            return newKnownTables;
+        }
+        // otherwise we cannot evaluate what tables, if any, are new.
+        return null;
```
I've actually changed my approach here a bit (or moved the logic around a bit) so this function no longer exists.

gunnarmorling commented:
Hey @mtagle, first off all, thanks a lot for working on this feature! I'm trying to take a closer look eventually, but I'm kinda overwhelmed by the size of the change. Is there any chance that you:

* provide some kind of high-level discussion here in this PR
* separate refactorings - not leading to changed behavior - from actual changes?

Ideally, we could integrate refactorings early on - overall I'm looking for ways to make all this a bit easier to digest. Any thoughts?

mtagle commented:
@gunnarmorling I'm working on a design document at present, so, hopefully when that's complete it will provide a guideline for what I plan on doing.  I completely understand that this is pretty overwhelming even at the moment!

The key thing that I really would appreciate an extra set of eyes on right now is actually just focusing on the mysql snapshot refactoring.  (That essentially covers the new SnapshotMySQLUtility class and the existing SnapshotServerReader.)  So while the diff isn't split into refactoring and new functionality, that would be a really helpful chunk that doesn't include any new functionality.

gunnarmorling commented on debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/SnapshotServerReader.java:
```diff
@@ -0,0 +1,199 @@
+ * 
+ * @author Randall Hauch
+ */
+public class SnapshotServerReader extends AbstractSnapshotReader {
```
What is a "server reader"? How does it differ from other implementations?

gunnarmorling commented on debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/SnapshotServerReader.java:
```diff
@@ -0,0 +1,199 @@
+     */
+    protected void execute() {
+        context.configureLoggingContext("snapshot");
+        final AtomicReference<String> sql = new AtomicReference<>();
```
What's the purpose of this reference?

gunnarmorling commented on debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/SnapshotTableReader.java:
```diff
@@ -0,0 +1,110 @@
+ * Class snapshotting an individual table
+ * todo use a chainedReader to link all these individual readers together? maybe? is that what I want?
+ * the chained reader will make sure all of an individual reader's sourceRecords are polled before starting a new one.
+ * is that what I want?
```
You tell me :) It's a bit of a general challenge for me that it's not quite clear to me where we're headed with this change.

gunnarmorling commented on debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/SnapshotMySQLUtility.java:
```diff
@@ -0,0 +1,600 @@
+    // todo buffer a given list of tables (those tables must be locked, of course)
+    // but that way we can lock all the tables but only buffer/read some tables.  Which is useful because then we
+    // can release the lock earlier than we could normally.
+    public void bufferLockedTables(List<TableId> tableIdsToBuffer,
```
A description of this method would be good. Also I think it still is very long, can we extract further things into separate methods?

gunnarmorling commented on debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/SnapshotMySQLUtility.java:
```diff
@@ -0,0 +1,600 @@
+                sql.set("USE " + quote(tableId.catalog()) + ";");
+                mysql.execute(sql.get());
+
+                AtomicLong numRows = new AtomicLong(-1);
```
E.g. this block, couldn't we extract something like `statementFactory = buildStatementFactory()` which would then encapsulate the entire logic for building the SF?

gunnarmorling commented on debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/SnapshotMySQLUtility.java:
```diff
@@ -0,0 +1,600 @@
+     * @throws SQLException
+     * @see {@link #getTableIds()}, {@link #getTableIdsByDbName()}, {@link #getReadableDatabaseNames()}
+     */
+    public void readTableList() throws SQLException {
```
A general remark on this class and is methods is that it seems to be based a lot on side effects. So e.g. this method seems to alter the state of this class instead of returning the list of tables. I feel this makes it tough to grok the overarching picture.

mtagle commented:
I'm closing this PR in favor of smaller piecemeal prs, at gunnar's suggestion.  I just put up one for just the snapshot refactor and closely related changes: https://github.com/debezium/debezium/pull/292 If you could reiterate any relevant, unaddressed comments there it would be appreciated!

uuhnaut69 commented:
@gunnarmorling I'm using embedded debezium , when binlog is expired, deleted and i have reprocessing event from beginning, how can i avoid it ?

gunnarmorling commented:
Tuan, this PR has been closed for more than two years and your
question seems not related to it. Please come to the mailing list
(https://groups.google.com/forum/?fromgroups=#!forum/debezium) for
asking questions like this. Thanks!
